### DISCUSSION TITLE
How to correctly get CG coefficients from e3nn.o3.wigner_3j

### DISCUSSION THREAD

**Discussion by smliu1997** [e3nn/e3nn#509] [Q&A]
I wonder how to correctly convert e3nn.o3.wigner_3j to CG coefficients? Here is the code I tried, and the output does not give the correct CG coefficients. Even the absolute values are incorrect. I followed the relation between wigner-3j and CG coefficients shown in wiki. 

```
import torch
from e3nn import o3
import math

# absolute values may be wrong

def wigner_3j_to_cg_v1(l1, l2, l3):
    w = o3.wigner_3j(l1, l2, l3)
    cg = torch.zeros((2 * l1 + 1, 2 * l2 + 1, 2 * l3 + 1))
    for m3 in range(-l3, l3 + 1):
        cg[:, :, l3 + m3] = (-1)**(-l1 + l2 - m3) * math.sqrt(2 * l3 + 1) * w[:, :, l3 - m3]
    return cg

def find_cg_v1(l1, l2, l3, m1, m2, m3):
    cg = wigner_3j_to_cg_v1(l1, l2, l3)
    assert -l1 <= m1 <= l1
    assert -l2 <= m2 <= l2
    assert -l3 <= m3 <= l3
    return cg[l1 + m1, l2 + m2, l3 + m3]

 # print computed value and expected value
print(find_cg_v1(2, 1, 3, 1, -1, 0), math.sqrt(0.2))
print(find_cg_v1(2, 1, 3, 0, 0, 0), math.sqrt(0.6))
print(find_cg_v1(2, 1, 3, -1, 1, 0), math.sqrt(0.2))

print(find_cg_v1(2, 1, 1, 1, -1, 0), math.sqrt(0.3))
print(find_cg_v1(2, 1, 1, 0, 0, 0), -math.sqrt(0.4))
print(find_cg_v1(2, 1, 1, -1, 1, 0), math.sqrt(0.3))



```

**Comment by smliu1997** [comment]
With more tests, I found this version gives correct absolute values (at least in the test set here), but signs may be wrong. 
```
import torch
from e3nn import o3
import math

# correct absolute values, but signs may be wrong

def wigner_3j_to_cg(l1, l2, l3):
    w = o3.wigner_3j(l1, l2, l3)
    cg = torch.zeros((2 * l1 + 1, 2 * l2 + 1, 2 * l3 + 1))
    for m1 in range(-l1, l1 + 1):
        cg[l1 + m1, :, :] = (-1)**(-l2 + l3 - m1) * math.sqrt(2 * l3 + 1) * w[l1 - m1, :, :]
    return cg

def find_cg(l1, l2, l3, m1, m2, m3):
    cg = wigner_3j_to_cg(l1, l2, l3)
    assert -l1 <= m1 <= l1
    assert -l2 <= m2 <= l2
    assert -l3 <= m3 <= l3
    return cg[l1 + m1, l2 + m2, l3 + m3]

 # print computed value and expected value
print(find_cg(2, 1, 3, 1, -1, 0), math.sqrt(0.2))
print(find_cg(2, 1, 3, 0, 0, 0), math.sqrt(0.6))
print(find_cg(2, 1, 3, -1, 1, 0), math.sqrt(0.2))

print(find_cg(2, 1, 1, 1, -1, 0), math.sqrt(0.3))
print(find_cg(2, 1, 1, 0, 0, 0), -math.sqrt(0.4))
print(find_cg(2, 1, 1, -1, 1, 0), math.sqrt(0.3))


```

**Reply by zx-ml** [reply]
Have you solved the problem? I have the same question. Is it possible that what o3.wigner_3j returns is cg coefficient, because in e3nn's paper the author says that "The Clebsch–Gordan coefficients (also called Wigner 3j symbols)"?

**Reply by mariogeiger** [reply]
The cg are not strictly defined, there is many different conventions.
